lib/compat/wordpress-6.8/blocks.php

```diff
@@ -6,40 +6,85 @@
  */
 
 /**
- * Filters the block type arguments during registration to stabilize experimental block supports.
+ * Filters the block type arguments during registration to stabilize
+ * experimental block supports.
  *
- * This is a temporary compatibility shim as the approach in core is for this to be handled
- * within the WP_Block_Type class rather than requiring a filter.
+ * This is a temporary compatibility shim as the approach in core is for this
+ * to be handled within the WP_Block_Type class rather than requiring a filter.
  *
  * @param array $args Array of arguments for registering a block type.
  * @return array Array of arguments for registering a block type.
  */
 function gutenberg_stabilize_experimental_block_supports( $args ) {
-	if ( empty( $args['supports']['typography'] ) ) {
+	if ( empty( $args['supports'] ) ) {
 		return $args;
 	}
 
-	$experimental_typography_supports_to_stable = array(
-		'__experimentalFontFamily'     => 'fontFamily',
-		'__experimentalFontStyle'      => 'fontStyle',
-		'__experimentalFontWeight'     => 'fontWeight',
-		'__experimentalLetterSpacing'  => 'letterSpacing',
-		'__experimentalTextDecoration' => 'textDecoration',
-		'__experimentalTextTransform'  => 'textTransform',
+	$experimental_to_stable_keys = array(
+		'typography'           => array(
+			'__experimentalFontFamily'     => 'fontFamily',
+			'__experimentalFontStyle'      => 'fontStyle',
+			'__experimentalFontWeight'     => 'fontWeight',
+			'__experimentalLetterSpacing'  => 'letterSpacing',
+			'__experimentalTextDecoration' => 'textDecoration',
+			'__experimentalTextTransform'  => 'textTransform',
+		),
+		'__experimentalBorder' => 'border',
 	);
 
-	$current_typography_supports = $args['supports']['typography'];
-	$stable_typography_supports  = array();
+	$updated_supports = array();
+	foreach ( $args['supports'] as $support => $config ) {
+		// Add the support's config as is when it's not in need of stabilization.
+		if ( empty( $experimental_to_stable_keys[ $support ] ) ) {
+			$updated_supports[ $support ] = $config;
+			continue;
+		}
+
+		// Stabilize the support's key if needed e.g. __experimentalBorder => border.
+		if ( is_string( $experimental_to_stable_keys[ $support ] ) ) {
+			$stabilized_key = $experimental_to_stable_keys[ $support ];
+
+			// If there is no stabilized key present, use the experimental config as is.
+			if ( ! array_key_exists( $stabilized_key, $args['supports'] ) ) {
+				$updated_supports[ $stabilized_key ] = $config;
+				continue;
+			}
+
+			// Determine the order of keys, so the last defined can be preferred.
```

Comment:
This could be just me, but should we add why the last defined is preferred? The logic looks good, but it code itself doesn't explain the reason.

I could be missing something, so feel free to ignore me. I'm only thinking out loud and wondering if folks looking at this code later might need the hint. 

> it's unlikely we can cover all permutations of filters and plugins that might update their use of experimental support keys at different rates.

Based on this, and what's been already explained to me, is it to increase the probability of accurately processing the last known support value, which could originate from late-running hooks?

Reply:
Yes. Filters could run in any order and we can't guarantee whether they'll use the experimental or stable key. 

This approach isn't bulletproof but should work for the vast majority of cases while becoming less and less of a possibility over time as extenders migrate to the stable key.

Some simple use cases this order-based approach tackles include:

1. Core blocks are updated to use stable keys but a plugin filters the block type args and sets a experimental key.
    - The expected outcome would be that the plugin's filtered value is use. 
    - If the block was already migrated to stable keys, the filtered experimental key would be added to the supports config after the stable key. 
    - Using the last defined key's value here allows the expected behaviour.
2. Take the flip side of the above scenario, the block uses an experimental key but a plugin filters things setting the stable key. 
    - Again as it was set later, then it should take precedence.

We get into trouble if multiple plugins all filter a blocks supports using a mix of experimental and stable keys one after another. There's no way we can control that scenario or determine with certainty which key's value should be used. The best we can do for now is use the insertion order for the keys. This is still better than simply using stable keys and only falling back to experimental keys in their absence.

Another option to mitigate a little of the risk here is to strongly recommend extenders update any block supports in their custom blocks or filters to use the stable keys ASAP. If a dev note isn't sufficient here, then a make post might be in order. 

All that aside, I'll try and improve that inline comment.

Reply:
I've had a crack at updating the inline comments in https://github.com/WordPress/gutenberg/pull/66918/commits/9c32c4ae41e6fb1bdae0044c622c9c2514f1f3fc. Let me know if you think that's adequate.

I'm happy to tweak further if there's a clearer, more eloquent, way of communicating this.

Reply:
I appreciate the extra comments, thank you!

Don't let my questions stop the merge. Everything is iterable! 🚢